<div class="s-prose s-prose__disable-spoiler-hover js-post-body" itemprop="text">
<p>I see one possible problem with this approach namely that any exception (except for a <code>ThreadAbortException</code>) which is maybe thrown by the <code>Func&lt;Task&gt; t</code> is silently swallowed if the mutex couldn't be freed.  </p>

<p>See: <a href="https://stackoverflow.com/a/2911229/2655508">what happens if a finally block throws an exception</a></p>

<p>A possible workaround could be to have a <code>catch</code> block which stores the exception in a variable which is later passed as an innerexception to the <code>new Exception</code>.  </p>

<hr>

<p>The naming of the <code>struct</code> and the method parameter is not very good, because it doesn't tell a reader of the code anything about it. Try to use better names.  </p>
    </div>